### PR TITLE
Remove use of the download lock and publish flag from the system

### LINKED ISSUE #191: Remove usage of the published flag from the system
Throughout our services, mainly the learning object service, the published flag is used in our queries to to pull objects from the database. This needs to change and query the status of an object based on a users access privileges.

### LINKED ISSUE #192: Add authorization to fetch released and submitted revisions based on user privileges.
Queries for learning objects by authorized users (reviewers, curators, editors, admins) should return both the released object and any submitted revisions.

Requires #199

### LINKED ISSUE #193: Remove download locks from the system
Once an object is released there shouldn't be a download lock on it. The use of the download lock has become obsolete since a learning object that is visible to the public must have downloadable content otherwise it is not a true 'released' object.

### PR DESCRIPTION
**Pull request by davidnso**
This PR removes the usage of the published flag and download lock. Additionally, this PR adds authorization on read operations in replace the usage of the published flag. This allows authorized users to read released and unreleased objects.

Closes #191
Closes #192
Closes #193 

Needs Cyber4All/clark-client#613
Needs Cyber4All/clark-client#617

### CODE REVIEW

File: src/drivers/MongoDriver.ts [hunk 1]
```diff
@@ -777,7 +778,7 @@ export class MongoDriver implements DataStore {
       full,
     );
 
-    if (!accessUnpublished && !learningObject.published)
+    if (!accessUnreleased && !learningObject.published)
```
Comment (nvisal1) [inline src/drivers/MongoDriver.ts]: Can you explain why the published flag is still being used here?

Comment (davidnso) [reply src/drivers/MongoDriver.ts]: This flag must've slipped through on this branch, it'll be removed on the next commit to check the learning objects status.

File: src/drivers/MongoDriver.ts [hunk 2]
```diff
@@ -789,23 +790,23 @@ export class MongoDriver implements DataStore {
...
+        query.status = RELEASED;
       }
 
       let objectCursor = await this.db
         .collection(COLLECTIONS.LEARNING_OBJECTS)
         .find<LearningObjectDocument>(query);
       const totalRecords = await objectCursor.count();
       objectCursor = this.applyCursorFilters(objectCursor, { page, limit });
-
+      console.log(query)
```
Comment (DonSeannelly) [inline src/drivers/MongoDriver.ts]: Looks like a log statement was left in here

File: src/interactors/LearningObjectInteractor.ts [hunk 3]
```diff
@@ -32,17 +32,17 @@ export class LearningObjectInteractor {
...
-      // Set accessUnpublished
-      let accessUnpublished = params.accessUnpublished;
+      // Set accessUnreleased
+      let accessUnreleased = params.accessUnreleased;
 
-      // If accessUnpublished is unset, set equal to the result of the hasOwnership function
-      if (accessUnpublished === undefined || accessUnpublished === null) {
-        accessUnpublished = await this.hasOwnership({
+      // If accessUnreleased is unset, set equal to the result of the hasOwnership function
+      if (accessUnreleased === undefined || accessUnreleased === null) {
```
Comment (DonSeannelly) [inline src/interactors/LearningObjectInteractor.ts]: This could be simplified by using the double equals expression below, so that type coercion is not used and both null/undefined will register as being a bottom value:

```suggestion
      if (accessUnreleased == null) {
```

File: src/interactors/LearningObjectInteractor.ts [hunk 4]
```diff
@@ -32,17 +32,17 @@ export class LearningObjectInteractor {
...
     userToken: UserToken;
-    accessUnpublished?: boolean;
+    accessUnreleased?: boolean;
     loadChildren?: boolean;
     query?: LearningObjectQuery;
   }): Promise<LearningObject[]> {
     try {
-      // Set accessUnpublished
-      let accessUnpublished = params.accessUnpublished;
+      // Set accessUnreleased
```
Comment (DonSeannelly) [inline src/interactors/LearningObjectInteractor.ts]: This comment doesn't seem to explain a lot. Unless there's a need to elaborate further, I think it's probably not needed.

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 5]
```diff
@@ -48,9 +48,6 @@ export async function addLearningObject(
       if (!author.emailVerified) {
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectInteractor.ts]: `LearningObject`'s publish/unpublish is still being used here.

File: src/drivers/MongoDriver.ts [hunk 6]
```diff
@@ -765,7 +766,7 @@ export class MongoDriver implements DataStore {
   async fetchLearningObject(
     id: string,
     full?: boolean,
-    accessUnpublished?: boolean,
+    accessUnreleased?: boolean,
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Since we're querying on `status` instead of `published` I would suggest this value be converted to a non-binary value to support all possible values of `status`.

File: src/drivers/MongoDriver.ts [hunk 7]
```diff
@@ -777,7 +778,7 @@ export class MongoDriver implements DataStore {
       full,
     );
 
-    if (!accessUnpublished && !learningObject.published)
+    if (!accessUnreleased && learningObject.status!==RELEASED)
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: This logic should be done inside of the interactor when authorizing. A request to the datastore should not be made to fetch a resource a user doesn't have access to unless a part of that resource is needed for authorization.

File: src/drivers/MongoDriver.ts [hunk 8]
```diff
@@ -789,23 +790,23 @@ export class MongoDriver implements DataStore {
    * @returns {Cursor<LearningObjectRecord>[]} cursor of literally all objects
    */
   async fetchAllObjects(
-    accessUnpublished?: boolean,
+    accessUnreleased?: boolean,
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Since we're querying on `status` instead of `published` I would suggest this value be converted to a non-binary value to support all possible values of `status`.

File: src/drivers/MongoDriver.ts [hunk 9]
```diff
@@ -854,13 +855,13 @@ export class MongoDriver implements DataStore {
   async fetchMultipleObjects(
     ids: string[],
     full?: boolean,
-    accessUnpublished?: boolean,
+    accessUnreleased?: boolean,
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Since we're querying on `status` instead of `published` I would suggest this value be converted to a non-binary value to support all possible values of `status`.

File: src/drivers/MongoDriver.ts [hunk 10]
```diff
@@ -901,7 +902,7 @@ export class MongoDriver implements DataStore {
     level: string[];
     standardOutcomeIDs: string[];
     text: string;
-    accessUnpublished?: boolean;
+    accessUnreleased?: boolean;
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Since we're querying on `status` instead of `published` I would suggest this value be converted to a non-binary value to support all possible values of `status`.

File: src/drivers/MongoDriver.ts [hunk 11]
```diff
@@ -1049,7 +1050,7 @@ export class MongoDriver implements DataStore {
    * @memberof MongoDriver
    */
   private buildSearchQuery(
-    accessUnpublished: boolean,
+    accessUnreleased: boolean,
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Since we're querying on `status` instead of `published` I would suggest this value be converted to a non-binary value to support all possible values of `status`.

File: src/drivers/MongoDriver.ts [hunk 12]
```diff
@@ -37,6 +37,7 @@ export enum COLLECTIONS {
   LO_COLLECTIONS = 'collections',
   MULTIPART_STATUSES = 'multipart-upload-statuses',
 }
+const RELEASED = 'released';
```
Comment (nrwinner) [inline src/drivers/MongoDriver.ts]: We should use the `Status` enum here instead of declaring a separate variable to maintain a single source-of-truth. `Status` can be used like `LearningObject.Status.RELEASED` in our current configuration.

File: src/drivers/express/ExpressRouteDriver.ts [hunk 13]
```diff
@@ -146,19 +146,19 @@ export class ExpressRouteDriver {
       .route('/learning-objects/:username/:learningObjectName')
       .get(async (req, res) => {
         try {
-          let accessUnpublished = false;
+          let accessUnreleased = false;
           const username = req.params.username;
           const cookie = req.cookies.presence;
           if (cookie) {
```
Comment (gshaw1997) [inline src/drivers/express/ExpressRouteDriver.ts]: This logic should be handled inside of the interactor method.

File: src/drivers/MongoDriver.ts (hunk 9 above)
Comment (davidnso) [reply src/drivers/MongoDriver.ts]: I'd suggest adding another value that holds all the possible values of status because, this flag seems to be used at the top level to check whether or not the user is able to access released or unreleased material. So if they can access unreleased material, my suggestion is, in that case we should utilize all the other values of status when making a query in the MongoDriver.

File: src/drivers/MongoDriver.ts (hunk 12 above)
Comment (davidnso) [reply src/drivers/MongoDriver.ts]: I wasn't aware that the different statuses were already enums in the learning object entity, this change will be made on the next commit. Thank you

File: src/drivers/MongoDriver.ts (hunk 7 above)
Comment (davidnso) [reply src/drivers/MongoDriver.ts]: In another branch the "fetchLearningObject" function points to the released collection and all of those objects have a status of released, so would this line still be necessary? The accessUnreleased check would happen in the interactor then.

File: src/interactors/LearningObjectInteractor.ts (hunk 4 above)
Comment (davidnso) [reply src/interactors/LearningObjectInteractor.ts]: Yea these comments aren't needed

File: src/LearningObjects/LearningObjectInteractor.ts (hunk 5 above)
Comment (gshaw1997) [reply src/LearningObjects/LearningObjectInteractor.ts]: The status should be `unreleased` if the user is not verified.

File: src/LearningObjectSubmission/SubmissionInteractor.ts [hunk 14]
```diff
@@ -57,20 +56,32 @@ export async function cancelSubmission(
  * @returns {void}
  */
 export async function createChangelog(params: {
-  dataStore: DataStore,
-  learningObjectId: string,
-  user: UserToken,
-  changelogText: string,
+  dataStore: DataStore;
```
Comment (DonSeannelly) [inline src/LearningObjectSubmission/SubmissionInteractor.ts]: I thought this change was odd, and ended up finding [this](https://github.com/prettier/prettier/issues/1944) about why `;` instead of `,`

File: src/interactors/AuthorizationManager.ts [hunk 15]
```diff
@@ -99,3 +139,54 @@ async function userIsOwner(params: { dataStore: DataStore; user: UserToken; obje
...
 }
 
+/**
+ * Checks if accessGroups contains admin or editor
+ *
+ * @param {string[]} accessGroups
+ * @returns {boolean}
+ */
+export function isAdminOrEditor(accessGroups: string[]): boolean {
+  if (
```
Comment (DonSeannelly) [inline src/interactors/AuthorizationManager.ts]: This function can be simplified to:

```
return accessGroups.includes(UserRole.ADMIN) || accessGroups.includes(UserRole.EDITOR);
```

File: src/interactors/LearningObjectInteractor.ts [hunk 16]
```diff
@@ -7,19 +7,50 @@ import {
...
   updateObjectLastModifiedDate,
   updateParentsDate,
 } from '../LearningObjects/LearningObjectInteractor';
+import { sanitizeText, sanitizeObject } from '../functions';
 
 // file size is in bytes
 const MAX_PACKAGEABLE_FILE_SIZE = 100000000;
 
+const LearningObjectState = {
+  UNRELEASED: [
```
Comment (DonSeannelly) [inline src/interactors/LearningObjectInteractor.ts]: This seems kind of confusing since unreleased/review/released are groups, as well as individual statuses. Not sure on the solution here. Maybe we need to define phases within the state transition diagram?

File: src/drivers/express/ExpressAdminRouteDriver.ts [hunk 17]
```diff
@@ -61,44 +54,15 @@ export class ExpressAdminRouteDriver {
...
-          res.sendStatus(200);
-        } catch (e) {
-          console.error(e);
-          if (e.includes('Invalid Access')) {
-            res.status(401).send(e);
-          } else {
-            res.status(500).send(e);
-          }
-        }
+        // Respond to clients that this functionality is now gone
```
Comment (nrwinner) [inline src/drivers/express/ExpressAdminRouteDriver.ts]: Thank you for this comment :heart:

File: src/interactors/AuthorizationManager.ts [hunk 18]
```diff
@@ -54,9 +79,13 @@ function hasPrivilegedAccess(user: UserToken, dataStore: DataStore, objectId: st
...
  */
-async function checkCollectionWriteAccess(params: { user: UserToken, dataStore: DataStore, objectId: string }): Promise<boolean> {
+async function checkCollectionWriteAccess(params: {
+  user: UserToken;
+  dataStore: DataStore;
+  objectId: string;
+}): Promise<boolean> {
   // Regex checks to see if the given objectId string contains an id or a name
-  const regexp = new RegExp('/^[a-f\d]{24}$/i');
+  const regexp = new RegExp('/^[a-fd]{24}$/i');
```
Comment (nvisal1) [inline src/interactors/AuthorizationManager.ts]: Why did you remove the \ ?

File: src/interactors/LearningObject.interactor.spec.ts [hunk 19]
```diff
@@ -1,48 +1,42 @@
...
+describe('loadUsersObjectSummaries', () => {
   it('should load learning object summary', done => {
-    return LearningObjectInteractor.loadLearningObjectSummary({
-      dataStore: driver,
+    return LearningObjectInteractor.loadUsersObjectSummaries({
+      dataStore,
       library,
       userToken: MOCK_OBJECTS.USERTOKEN,
-      username: MOCK_OBJECTS.USERNAME,
+      username: MOCK_OBJECTS.USERTOKEN.username,
```
Comment (nvisal1) [inline src/interactors/LearningObject.interactor.spec.ts]: Good change

File: src/interactors/LearningObjectInteractor.ts (hunk 16 above)
Comment (nrwinner) [reply src/interactors/LearningObjectInteractor.ts]: @DonSeannelly Can you elaborate on the source of confusion? What do you mean by "groups"?

File: src/interactors/LearningObject.interactor.spec.ts [hunk 20]
```diff
@@ -1,48 +1,42 @@
...
     })
       .then(val => {
         expect(val).to.be.an('array');
         done();
       })
       .catch(error => {
+        console.log('ERR: ', error);
```
Comment (nvisal1) [inline src/interactors/LearningObject.interactor.spec.ts]: Why are we logging the failure of a unit test?

File: src/LearningObjects/LearningObjectInteractor.ts [hunk 21]
```diff
@@ -166,26 +163,34 @@ export async function updateLearningObject(params: {
...
+        updates,
+      });
+      updates.date = Date.now().toString();
+      await params.dataStore.editLearningObject({
+        id: params.id,
+        updates,
+      });
+    } else {
+      return Promise.reject(
+        new Error('User does not have authorization to perform this action'),
```
Comment (nvisal1) [inline src/LearningObjects/LearningObjectInteractor.ts]: We should add this error to the LearningObjectError object and reference it.

File: src/LearningObjectSubmission/SubmissionInteractor.ts [hunk 22]
```diff
@@ -57,20 +56,32 @@ export async function cancelSubmission(
...
 }): Promise<void> {
   try {
-    const hasAccess = hasLearningObjectWriteAccess(params.user, params.dataStore, params.learningObjectId);
+    const hasAccess = hasLearningObjectWriteAccess(
+      params.user,
+      params.dataStore,
+      params.learningObjectId,
+    );
     if (hasAccess) {
       const authorID = await params.dataStore.findUser(params.user.username);
```
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/LearningObjectSubmission/SubmissionInteractor.ts]: This is out of scope of the PR as no code was changed here, just formatting. I'll go ahead and update to streamline to changes.

File: src/interactors/AuthorizationManager.ts (hunk 18 above)
Comment (gshaw1997) [reply src/interactors/AuthorizationManager.ts]: This was removed by Prettier, meaning it wasn't formatted properly. Was the intention to match digits?

File: src/interactors/LearningObjectInteractor.ts (hunk 16 above)
Comment (gshaw1997) [reply src/interactors/LearningObjectInteractor.ts]: What part is confusing? Is it the grouping of state of the naming used?

Comment (DonSeannelly) [reply src/interactors/LearningObjectInteractor.ts]: @nrwinner @gshaw1997 For instance, there is an object called `LearningObjectState`, which contains an array called `UNRELEASED`, which contains a value in it called `LearningObject.Status.UNRELEASED`.

I don't think that this should hold up this PR, its more of a discussion point we might want to bring up. If we do need to group certain Learning Object statuses together due to some logical meaning, which this PR seems to suggest, I think we should standardize a name for those groupings. 

Right now, in the way that it is presented here, I think that an array having the same name as one of it's members (albeit a different namespace) could be confusing.

### TL;DR 
If we mention an "unreleased" Learning Object in conversation, do we mean a LO with status `unreleased` or a status of either `unreleased`/`rejected`?
